internal/rbd/rbd_util.go

```diff
@@ -210,6 +213,85 @@ var supportedFeatures = map[string]imageFeature{
 	},
 }
 
+// getKrbdSupportedFeatures load the module if needed and return supported
+// features attribute as a string.
+func getKrbdSupportedFeatures() (string, error) {
+	// check if the module is loaded or compiled in
+	_, err := os.Stat(krbdSupportedFeaturesFile)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			log.ErrorLogMsg("stat on: %s failed, err: %v", krbdSupportedFeaturesFile, err)
+
+			return "", err
+		}
+		// try to load the module
+		_, _, err = util.ExecCommand(context.TODO(), "modprobe", rbdDefaultMounter)
+		if err != nil {
+			log.WarningLogMsg("modprobe failed with error %v", err)
+
+			return "", err
+		}
+	}
+	val, err := ioutil.ReadFile(krbdSupportedFeaturesFile)
+	if err != nil {
+		log.ErrorLogMsg("ReadFile %s failed, err: %v", krbdSupportedFeaturesFile, err)
+
+		return "", err
+	}
+
+	return strings.TrimSuffix(string(val), "\n"), nil
+}
+
+// hexStringToInteger convert hex value to uint.
+func hexStringToInteger(hexString string) (uint64, error) {
+	// replace 0x or 0X with empty String
+	numberStr := strings.Replace(hexString, "0x", "", -1)
+	numberStr = strings.Replace(numberStr, "0X", "", -1)
+
+	output, err := strconv.ParseUint(numberStr, 16, 64)
+	if err != nil {
+		log.ErrorLogMsg("strconv.ParseUint(%s) failed, err: %v", numberStr, err)
+
+		return 0, err
+	}
+
+	return output, nil
+}
+
+// isKrbdFeatureSupported checks if a given Image Feature is supported by krbd
+// driver or not.
+func isKrbdFeatureSupported(fname string) (bool, error) {
+	supported := false
+	attr, err := getKrbdSupportedFeatures()
+	if err != nil {
+		return supported, err
+	}
+
+	features, err := hexStringToInteger(attr)
+	if err != nil {
+		return supported, err
+	}
+
+	if strings.EqualFold(fname, librbd.FeatureNameLayering) {
```

Comment:
Better not hard-code all features here, let go-ceph handle that with the `FeatureSet` object? Then use `fs.Names()` to get a list of all supported features to match against?

Reply:
I had switched to switch case now. I will check on `FeatureSet`, thanks!